src/redisearch_rs/encode_decode/src/qint.rs

```diff
@@ -0,0 +1,284 @@
+///
+/// # Arguments
+/// * `cursor` - Buffer writer
+/// * `a` - First integer
+/// * `b` - Second integer
+/// * `c` - Third integer
+///
+/// # Returns
+/// The number of bytes written to the buffer or an io error
+pub fn qint_encode3<W>(cursor: &mut W, a: u32, b: u32, c: u32) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    qint_encode(cursor, [a, b, c])
+}
+
+/// Encodes 2 integers into a QInt buffer
+///
+/// # Arguments
+/// * `cursor` - Buffer writer
+/// * `a` - First integer
+/// * `b` - Second integer
+///
+/// # Returns
+/// The number of bytes written to the buffer or an io error
+pub fn qint_encode2<W>(cursor: &mut W, a: u32, b: u32) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    qint_encode(cursor, [a, b])
+}
+
+// Internal: Encodes one byte of using qint encoding, called in a loop.
+#[inline(always)]
+fn qint_encode_stepwise<W>(
+    leading: &mut u8,
+    cursor: &mut W,
+    mut value: u32,
+    offset: u32,
+) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    let mut ret: usize = 0;
+    let mut num_bytes: i8 = -1;
+    loop {
+        cursor.write_all(&[value as u8])?;
+        ret += &[value as u8].len();
+        num_bytes += 1;
+
+        // shift right until we have no more bigger bytes that are non zero
+        value >>= 8;
+        // do while(value) in c
+        if value == 0 {
+            break;
+        }
+    }
+    // encode the bit length of our integer into the leading byte.
+    // 0 means 1 byte, 1 - 2 bytes, 2 - 3 bytes, 3 - 4 bytes.
+    // we encode it at the i*2th place in the leading byte
+    *leading |= (num_bytes as u8) << (offset * 2);
+    Ok(ret)
+}
+
+/// Internal: Decode an integer value from a buffer based on bit width
+///
+/// # Parameters
+/// * `bit_value` - The number of bits decoded from leading byte (0=1 byte, 1=2 bytes, 2=3 bytes, 3+=4 bytes)
+/// * `reader` - The buffer reader
+///
+/// # Returns
+/// A tuple containing (decoded_value, bytes_used)
+#[inline(always)]
+fn qint_decode_value<R>(bit_value: u8, reader: &mut R) -> Result<(u32, usize), std::io::Error>
+where
+    R: Read,
+{
+    match bit_value {
+        0 => {
+            // 1 byte
+            let mut buf = [0; 1];
+            reader.read_exact(&mut buf)?;
+            Ok((buf[0] as u32, 1))
+        }
+        1 => {
+            // 2 bytes
+            let mut buf = [0; 2];
+            reader.read_exact(&mut buf)?;
+            Ok((u16::from_ne_bytes(buf) as u32, 2))
+        }
+        2 => {
+            // 3 bytes (mask off highest byte)
+            let mut buf = [0; 3];
+            reader.read_exact(&mut buf)?;
+            let bytes = [buf[0], buf[1], buf[2], 0];
+            Ok((u32::from_ne_bytes(bytes), 3))
+        }
+        _ => {
+            // 4 bytes
+            let mut buf = [0; 4];
+            reader.read_exact(&mut buf)?;
+            let bytes = [buf[0], buf[1], buf[2], buf[3]];
+            Ok((u32::from_ne_bytes(bytes), 4))
+        }
```

Comment:
We should split this into a branch for the 4 bytes case and an `unreachable!` branch.